stew/results.nim

```diff
@@ -542,111 +770,84 @@ func error*[T, E](self: Result[T, E]): E =
       raiseResultDefect("Trying to access error when value is set", self.v)
     else:
       raiseResultDefect("Trying to access error when value is set")
-  self.e
-
-template value*[T, E](self: Result[T, E]): T =
-  mixin get
-  self.get()
-
-template value*[T, E](self: var Result[T, E]): T =
-  mixin get
-  self.get()
-
-template valueOr*[T, E](self: Result[T, E], def: T): T =
+  when E isnot void:
+    self.e
+
+func tryError*[T, E](self: Result[T, E]): E {.inline.} =
+  ## Fetch error of result if set, or raise
+  ## Raises a ResultError[T]
+  mixin raiseResultOk
+  if self.o: self.raiseResultOk()
+  when E isnot void:
+    self.e
+
+# Alternative spellings for get
+template value*[T, E](self: Result[T, E]): T = self.get()
+template value*[T: not void, E](self: var Result[T, E]): var T = self.get()
+
+template valueOr*[T: not void, E](self: Result[T, E], def: T): T =
```

Comment:
Still suffers from double evaluation of `self`

Reply:
9bcba3ade870c30d92e4614461380a56f94179a6